File: algorithms/[27]移除元素/solution.ts
// topics = ["数组", "双指针"]

function removeElement(nums: number[], val: number): number {
  let i = 0;
  let j = nums.length;

  while (i < j) {
    if (nums[i] == val) {
      nums[i] = nums[j - 1];
      j -= 1;
    } else {
      i += 1;
    }
  }

  return i;
}

File: algorithms/[27]移除元素/solution.py
# topics = ["数组", "双指针"]

from typing import List


class Solution:
    def removeElement(self, nums: List[int], val: int) -> int:
        n = len(nums)
        i, j = 0, n - 1

        while True:
            while i < n and nums[i] != val:
                i += 1
            while j >= 0 and nums[j] == val:
                j -= 1

            if i > j:
                return i

            nums[i], nums[j] = nums[j], nums[i]
